### PR TITLE
Tuple field access using array index operator

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
The compiler already has an implementation allowing field access in tuples, using the syntax `t.f0`.
This is a new feature slated for P4-16 1.3.0.
The language-design working group proposed an alternative syntax, using `t[0]`. 
This PR is an implementation using the new syntax.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
@smolkaj The bazel build seems to have failed for some reasons that do not seem connected to the code changes.

**Comment by mbudiu-vmw** [comment]
I have marked this as draft until I sort out the CI errors.

**Comment by mbudiu-vmw** [comment]
Most of the errors come because we are now more strict about checking whether an empty struct is uninitialized.
To avoid warnings about an empty struct you still have to initialize it with an empty list expression.

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 1]
```diff
@@ -2005,47 +2005,63 @@ const IR::Node* TypeInference::postorder(IR::ArrayIndex* expression) {
...
             return expression;
         }
-        int index = cst->asInt();
+        index = cst->asInt();
         if (index < 0) {
             typeError("%1%: Negative array index %2%", expression, cst);
             return expression;
         }
+    }
+    // if index is negative here it means it's not a constant
```
Comment (jnfoster) [inline frontends/p4/typeChecking/typeChecker.cpp:2020]: Not a huge fan of this... isn't there a cleaner way to encode this analysis?

**Review by jnfoster — APPROVED** [review]

**Pull request merged**